unlock-app/src/components/creator/CreatorLockForm.js

```diff
@@ -14,86 +14,157 @@ import {
+      const invalidError = this.validate(field, state[field])
+      fieldValidity[field] = !invalidError
+      if (!invalidError) {
+        return fieldValidity
+      }
+      errors.push(invalidError)
+      return fieldValidity
+    }, {})
+
+    // the form can be submitted if and only if there are no errors triggered by any field
+    validityState.formValid = errors.length === 0
+    return { validityState, errors }
   }
 
   validate(name, value) {
-    let { unlimitedKeys } = this.state
-    return (
-      (name === 'name' && typeof value === 'string' && value.length > 0) ||
-      ((name === 'expirationDuration' || name === 'maxNumberOfKeys') &&
-        !isNaN(value) &&
-        value > 0) ||
-      (name === 'maxNumberOfKeys' && value === '∞' && unlimitedKeys) ||
-      (name === 'keyPrice' && !isNaN(value) && value >= 0)
-    )
+    const missing = val => val === undefined || val === null || val === ''
+    const isPositiveInteger = val =>
+      !isNaN(val) && +val === parseInt(val) && +val > 0
+    const isPositiveNumber = val => !isNaN(val) && +val > 0
+    switch (name) {
+      case 'name':
+        if (missing(value)) {
+          return FORM_LOCK_NAME_MISSING
+        }
+        break
+      case 'expirationDuration':
+      case 'maxNumberOfKeys': {
+        const error = {
+          expirationDuration: FORM_EXPIRATION_DURATION_INVALID,
+          maxNumberOfKeys: FORM_MAX_KEYS_INVALID,
+        }[name]
+        if (value !== '∞' && !isPositiveInteger(value)) {
+          return error
+        }
+        break
+      }
+      case 'keyPrice':
+        if (!isPositiveNumber(value)) {
+          return FORM_KEY_PRICE_INVALID
+        }
+        break
+    }
+    return false
   }
 
   handleUnlimitedClick() {
-    this.setState({ unlimitedKeys: true, maxNumberOfKeys: '∞' })
-    // Marking the maxNumberOfKeysRef field as valid.
-    this.maxNumberOfKeysRef.current.dataset.valid = true
+    this.setState(state => ({
+      unlimitedKeys: true,
+      maxNumberOfKeys: '∞',
+      valid: this.valid({ ...state, [name]: '∞' }),
+    }))
   }
 
-  handleChange(event) {
-    let { name, value, dataset } = event.target
-
-    if (name === 'maxNumberOfKeys') {
-      dataset.valid = value === '∞' ? true : this.validate(name, value)
-      this.setState({ unlimitedKeys: value === '∞' })
-    } else {
-      dataset.valid = this.validate(name, value)
-    }
-
-    this.setState({ [name]: value })
+  handleChange({ target: { name, value } }) {
+    this.setState(state => ({
+      unlimitedKeys:
+        name === 'maxNumberOfKeys' ? value === '∞' : state.unlimitedKeys,
+      [name]: value,
+      valid: this.valid({ ...state.valid, [name]: value }),
+    }))
   }
 
   handleSubmit() {
-    if (document.querySelector('[data-valid="false"]')) return false
+    this.setState(state => {
+      const { validityState: valid, errors } = this.valid(state)
+      const { setError, resetError } = this.props
```

Comment:
Pull that out of the `setState` callback (I think we should do the assignments at the beginning of the function)

Reply:
the only issue is that setState callbacks are async, I am not sure if we can rely on props in handleSubmit being the same when we finish. Although it may be OK, because it doesn't save any processor cycles or add other benefits, I don't think we should take this risk. What do you think?